### DISCUSSION TITLE
How to extend User model in MongoDB / Typescript?

### DISCUSSION THREAD

**Discussion by fsa317** [nextauthjs/next-auth#7217] [Help]
I want to add a few additional fields to my user object, they don't necessarily come from the oauth provider, they are app specific.  First question, should I extend user vs. create a separate model "AppUserInfo" that joins with the user collection?

If extending user makes the most sense, my question is how to do it?  I see lots of different things (using profile callback, extending TypeORM but they all feel incomplete).  Can anyone outline the steps for extending User for Mongo (using TS).

**Comment by 808vita** [comment]
I think you would be required to use the mongodb adapter  ?

**Reply by mike14747** [reply]
I'm using the native mongodb driver and next-auth CredentialsProvider. I add properties to the session and token this way...

I add this file to extend the user to support the additional properties:

```ts
// /types/next-auth.d.ts

import { DefaultSession } from 'next-auth';

declare module 'next-auth' {
    interface User {
        role: string;
    }

    interface Session extends DefaultSession {
        user?: User;
    }
}

declare module 'next-auth/jwt' {
    interface JWT {
        id: string;
        name: string;
        role: string;
    }
}
```

Make sure to add the /types folder to your tsconfig.ts.

```ts
"typeRoots": [
    "./types",
    "./node_modules/@types"
],
```

Then, in my [...nextauth]/route.ts callbacks, I add the properties.

```ts
callbacks: {
    async jwt({ token, user }) {
        if (user?.id) token.id = user.id;
        if (user?.name) token.name = user.name;
        if (user?.role) token.role = user.role;
        return token;
    },
    async session({ session, token }) {
        if (token.id && session.user) session.user.id = token.id;
        if (token.role && session.user) session.user.role = token.role;
        return session;
    },
},
```

**Reply by fsa317** [reply]
This looks close to what I am trying to do.  Two questions:

1. How do you force the use of a JWT token?  I tried to configure it with the Mongo adapter but no luck.
2. Where would a "lookup" go that looks up additional attributes from the DB?  My goal is to only look these up once per session.

**Reply by 808vita** [reply]
Will attach links to files relevant  to this , maybe looking at working implementation will help you .

https://github.com/808vita/nextjs-blog-with-auth/blob/deploy/src/pages/api/auth/%5B...nextauth%5D.ts

https://github.com/808vita/nextjs-blog-with-auth/tree/deploy/db
(folder containing related db code files)

https://github.com/808vita/nextjs-blog-with-auth
(link to root of the repo)

Ps. going to sleep soon , but I will try to respond back after waking up .
